client/state/data-layer/wpcom-http/pipeline/README.md

```diff
@@ -0,0 +1,98 @@
+The initial entry into the HTTP system hits the pipeline ingress.
+The final exit from the the HTTP system hits the pipeline egress.
+
+### Ingress
+
+Requests on ingress run through the pipeline as if it were some kind of funnel or chain.
+At this stage we can perform the following kinds of operations on the requests:
+ - filter out and drop unnecessary or redundant requests
+ - transform them by manipulating their data
+ - queue them for later if offline
+ - persist them to storage
+
+### Egress
+
+Requests on egress run through the pipeline to provide a way to follow-up with requests previously sent out.
+At this stage we can perform the following kinds of operations on the requests:
+ - Clean up and close out operations started on ingress
+ - Audit and perform accounting on requests as they come back (for example, track failed requests)
+
+## Detailed Operation
+
+To understand how this layer works we should start by examining the normal flow of HTTP requests and then see how we can augment that flow by inserting optimizations or processes into the pipeline.
+
+### The basic HTTP flow without the pipeline
+<!-- the following diagram was generated in draw.io - it can be edited by pasting in the contents of the SVG itself -->
+![Basic HTTP flow](https://cldup.com/X4mRbNKSaC.svg)
+
+#### Step A: HTTP request actions are dispatched
+
+Redux middleware can dispatch certain actions which describe HTTP requests to the WordPress.com API.
+These actions include information such as the API path, query arguments, request body, etc…
+The actions flow through the normal Redux dispatch chain and can be inspected with the Redux tools.
+
+#### Step B: HTTP layer intercepts requests
+
+The HTTP middleware handler intercepts the request actions and performs the actual network requests and response-handling.
+It is the piece that ties together outbound requests with the follow-up actions on upload progress, success, and failure events.
+
+#### Step C: Request is sent outbound
+
+When a request goes out "on the wire" there is an implicit tracking of that request in what we could consider a queue.
+At the time of this writing this tracking takes place within a `Promise` and its `then()` and `catch()` methods.
+
+#### Step D: Request returns or fails
+
+Requests come back in and are re-associated with the action that called for them.
+
+#### Step E: Follow-up actions are dispatched
+
+If the original action contained an `onSuccess` or `onFailure` responder (also a Redux action) then that will be dispatched in response according to the success or failure state of the network request.
+
+### HTTP flow _with_ the pipeline
+
+The pipeline injects itself at the ingress and egress of the actual network operations.
+It is the gatekeeper and the overseer.
+
+These gates are designed to introduce augmentation, introspection, and optimization into the HTTP layer which are _specifically_ and _only_ related to HTTP requests.
```

Comment:
we can insert a processor which could do logging or auditing of some kind.

I have a pet project I want to implement as soon as this layer gets out there whichever way: record the response times and content size for all network requests in a Calypso session and then export the data for analysis